src/io/simulator/simulator_handle.hpp

```diff
@@ -0,0 +1,453 @@
+        is_awaited_([](void *ptr) { return static_cast<ResponsePromise<T> *>(ptr)->IsAwaited(); }),
+        fill_([](void *this_ptr, OpaqueMessage opaque_message) {
+          T message = std::any_cast<T>(std::move(opaque_message.message));
+          auto response_envelope = ResponseEnvelope<T>{.message = std::move(message),
+                                                       .request_id = opaque_message.request_id,
+                                                       .from_address = opaque_message.from_address};
+          ResponsePromise<T> *promise = static_cast<ResponsePromise<T> *>(this_ptr);
+          auto unique_promise = std::unique_ptr<ResponsePromise<T>>(promise);
+          unique_promise->Fill(std::move(response_envelope));
+        }),
+        time_out_([](void *ptr) {
+          ResponsePromise<T> *promise = static_cast<ResponsePromise<T> *>(ptr);
+          auto unique_promise = std::unique_ptr<ResponsePromise<T>>(promise);
+          ResponseResult<T> result = TimedOut{};
+          unique_promise->Fill(std::move(result));
+        }) {}
+
+  bool IsAwaited() {
+    MG_ASSERT(ptr_ != nullptr);
+    return is_awaited_(ptr_);
+  }
+
+  void TimeOut() {
+    MG_ASSERT(ptr_ != nullptr);
+    time_out_(ptr_);
+    ptr_ = nullptr;
+  }
+
+  void Fill(OpaqueMessage &&opaque_message) {
+    MG_ASSERT(ptr_ != nullptr);
+    fill_(ptr_, std::move(opaque_message));
+    ptr_ = nullptr;
+  }
+
+  ~OpaquePromise() {
+    if (nullptr != ptr_) {
+      dtor_(ptr_);
+    }
+  }
+};
+
+struct DeadlineAndOpaquePromise {
+  uint64_t deadline;
+  OpaquePromise promise;
+};
+
+class SimulatorHandle {
+  std::mutex mu_{};
+  std::condition_variable cv_;
+
+  // messages that have not yet been scheduled or dropped
+  std::vector<std::pair<Address, OpaqueMessage>> in_flight_;
+
+  // the responsese to requests that are being waited on
+  std::map<PromiseKey, DeadlineAndOpaquePromise> promises_;
+
+  // messages that are sent to servers that may later receive them
+  std::map<Address, std::vector<OpaqueMessage>> can_receive_;
+
+  uint64_t cluster_wide_time_microseconds_;
+  bool should_shut_down_ = false;
+  SimulatorStats stats_;
+  size_t blocked_on_receive_ = 0;
+  std::set<Address> server_addresses_;
+  std::mt19937 rng_;
+  SimulatorConfig config_;
+
+ public:
+  explicit SimulatorHandle(SimulatorConfig config)
+      : cluster_wide_time_microseconds_(config.start_time), rng_(config.rng_seed), config_(config) {}
+
+  void IncrementServerCountAndWaitForQuiescentState(Address address) {
+    std::unique_lock<std::mutex> lock(mu_);
+    server_addresses_.insert(address);
+
+    while (true) {
+      size_t blocked_servers = blocked_on_receive_;
+
+      for (auto &[promise_key, opaque_promise] : promises_) {
+        if (opaque_promise.promise.IsAwaited()) {
+          if (server_addresses_.contains(promise_key.requester_address)) {
+            blocked_servers++;
+          }
+        }
+      }
```

Comment:
Maybe extract this logic to a member function because of two reasons:
1. It can be reused in `MaybeTickSimulator`
2. You can give it a nice name to explain its meaning: `GetBlockedServersOnWait()`, if I am not mistaken.

Reply:
You missed this also.

Reply:
thanks, extracted and pushed